### ISSUE TITLE
Nelson modules and package management

### ISSUE THREAD

**Issue by JohanMabille** [Nelson-numerical-software/nelson-minimalist-core#7]
Hi,

Thanks for this amazing project! We are considering packaging nelson for [conda-forge](https://conda-forge.org/), and I would like to know if the additional modules of Nelson (i.e. not in this core repo) can be packaged independently. My intuition is that it should be possible if installing a Nelson module consists in copying some files at the right place so that Nelson can find it, but I haven't diven into the details and I wanted to check there was no additional step required.

**Comment by Nelson-numerical-software** [comment]
Hi,

Nelson is not in version 1.0.0
It is again time to change some things.
About modules, have you see: https://github.com/Nelson-numerical-software/module_skeleton
For some basic macro .m, 'addpath' not implemented in this minimalist core could do the job.

The main trouble is the definition of the 'core' of Nelson

**Comment by Nelson-numerical-software** [comment]
For me, the minimal core for Nelson requires more dependencies (blas, lapack, openmp, ffi, ...)
You can build it using:
```
cmake -DCMAKE_BUILD_TYPE=Release -DMINIMAL_BUILD=ON -G "Unix Makefiles" .
cmake --build . -- -j $(nproc)
```
https://github.com/Nelson-numerical-software/nelson/actions/runs/3730347761/jobs/6327297671

**Comment by JohanMabille** [comment]
> The main trouble is the definition of the 'core' of Nelson

I agree this might not be easy, however this is not a blocker for packaging Nelson, nor the fact that it is not 1.0 yet (we've been packaging beta for a while, it took almost 20 years to SciPy to go to 1.0, etc ;))

> For me, the minimal core for Nelson requires more dependencies (blas, lapack, openmp, ffi, ...)

The advantage of this repo is that it does not depend on any Fortran, which will make it easier to build for WASM. We can also make a nelson meta package on conda-forge that would install this minimal repo along with all dependencies from the minima build of the main repo that are missing from this one.

**Comment by Nelson-numerical-software** [comment]
We can try to rethink "format" for the modules external and internal using the current standards but we must keep in mind the "philosophy" of the m language.

Some basic questions/answers:

- load the content of a directory should be feasable: 
Currently, nelson can already load  ".m" (function) ".nex"  (C/C++ builtin)  files added by 'addpath'. 
it can be generalized (it is in fact dynamic library renamed as {"nexglx", "nexa64", "nexmaci", "nexmaci64", "nexmacm1", "nexw32", "nexw64" })
- how to set priority easily between modules (example webtools module needs json module)
currently for external module, nelson read a module.json (similar to package.json)
https://github.com/Nelson-numerical-software/module_skeleton/blob/master/module.json
https://nelson-9.gitbook.io/nelson/en/modules_manager/nmm
- how to set dependencies between version ...
npm and node_modules is a hell ;)

These changes requires some important modifications.
This must  be common with main nelson repository (breaking changes) but it is an interesting task for the future of Nelson :)

**Comment by Nelson-numerical-software** [comment]
minimalist-core does not contains any function or builtin.
Maybe, we can try to define a list of builtin required in the minimalist-core.
firsts could be 'addpath' currently available in functions_manager.
(I can try to do this, after 1st january, current todo list is full for this month, sorry)

**Comment by Nelson-numerical-software** [comment]
about 'addpath', I create another ticket
https://github.com/Nelson-numerical-software/nelson-minimalist-core/issues/11

**Comment by JohanMabille** [comment]
> minimalist-core does not contains any function or builtin. Maybe, we can try to define a list of builtin required in the minimalist-core. firsts could be 'addpath' currently available in functions_manager. (I can try to do this, after 1st january, current todo list is full for this month, sorry)

No worries and no hurry, we start to "play" with nelson and try to make a xeus kernel with it, we can iterate on the modules' list of the minimalist core when we have something that works.

**Comment by Nelson-numerical-software** [comment]
> > minimalist-core does not contains any function or builtin. Maybe, we can try to define a list of builtin required in the minimalist-core. firsts could be 'addpath' currently available in functions_manager. (I can try to do this, after 1st january, current todo list is full for this month, sorry)
> 
> Thanks! And no hurry, we start to "play" with nelson and try to make a xeus kernel with it, we can iterate on the modules' list of the minimalist core when we have something that works.

No so complicated finally to add 'addpath' see #11 PR will come quickly

**Comment by Nelson-numerical-software** [comment]
'addpath' merged

**Comment by JohanMabille** [comment]
Thanks a lot! 
Happy new year ;)

**Comment by Nelson-numerical-software** [comment]
We import some builtin and functions (.m maybe also, currently no here)
Expected list (no new dependency required) modules:
- types
- string
- single
- overload
- logical
- integer
- sparse (need to be investigate if eigen is enough)
- handle (basic type)
- function handle
- error_manager
- elementary_functions (partially)
- constructors
- time could be also interesting but need to change code (boost -> c++17)
